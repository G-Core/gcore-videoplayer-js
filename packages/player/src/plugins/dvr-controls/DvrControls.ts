import { Events, Playback, UICorePlugin, template } from '@clappr/core'
import assert from 'assert'
import { trace } from '@gcorevideo/utils'

import { CLAPPR_VERSION } from '../../build.js'

import dvrHTML from '../../../assets/dvr-controls/index.ejs'
import '../../../assets/dvr-controls/dvr_controls.scss'
import { MediaControl } from '../media-control/MediaControl.js'

const T = 'plugins.dvr_controls'

/**
 * `PLUGIN` that adds the DVR controls to the media control UI
 *
 * @beta
 *
 * @remarks
 * Depends on:
 *
 * - {@link MediaControl}
 *
 * The plugin renders live stream indicator.
 * If DVR is enabled, the indicator shows whether the current position is at the live edge of the stream or not.
 * In the latter case, the indicator can be clicked to seek to the live edge.
 */
export class DvrControls extends UICorePlugin {
  private static readonly template = template(dvrHTML)

  /**
   * @internal
   */
  get name() {
    return 'dvr_controls'
  }

  /**
   * @internal
   */
  get supportedVersion() {
    return { min: CLAPPR_VERSION }
  }

  /**
   * @internal
   */
  override get events() {
    return {
      'click #gplayer-mc-back-to-live': 'clicked',
    }
  }

  /**
   * @internal
   */
  override get attributes() {
    return {
      class: 'dvr-controls',
      'data-dvr': '',
    }
  }

  /**
   * @internal
   */
  override bindEvents() {
    this.listenToOnce(this.core, Events.CORE_READY, this.onCoreReady)
    this.listenTo(
      this.core,
      Events.CORE_ACTIVE_CONTAINER_CHANGED,
      this.onActiveContainerChanged,
    )
  }

  private onCoreReady() {
    const mediaControl = this.core.getPlugin('media_control')
    assert(mediaControl, 'media_control plugin is required')

    this.listenTo(mediaControl, Events.MEDIACONTROL_RENDERED, this.mount)
  }

  private onActiveContainerChanged() {
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_LOADEDMETADATA,
      this.onMetadataLoaded,
    )
    this.listenTo(
      this.core.activeContainer,
      Events.CONTAINER_PLAYBACKDVRSTATECHANGED,
      this.onDvrStateChanged,
    )
  }

  private clicked() {
    const container = this.core.activeContainer
    if (!container.isPlaying()) {
      container.play()
    }
    container.seek(container.getDuration())
  }

  /**
   * @internal
   */
  override render() {
    trace(`${T} render`)
    this.$el.html(
      DvrControls.template({
        i18n: this.core.i18n,
      }),
    )
    this.$el.find('#gplayer-mc-back-to-live').hide()
    this.$el.find('#gplayer-mc-live').hide()

    return this
  }

  private onMetadataLoaded() {
    this.mount()
    this.toggleState(this.core.activeContainer.isDvrInUse())
  }

  private mount() {
    // TODO move mount point management logic to MediaControl
    if (this.core.getPlaybackType() !== Playback.LIVE) {
      return
    }
    const mediaControl = this.core.getPlugin('media_control') as MediaControl
    assert(mediaControl, 'media_control plugin is required')
    // TODO -> to MediaControl (auto hide)
    mediaControl.toggleElement('duration', false)
    mediaControl.toggleElement('position', false)
    mediaControl.mount('left', this.$el) // TODO use independent mount point
  }

  private onDvrStateChanged(dvrInUse: boolean) {
    this.toggleState(dvrInUse)
  }

  private toggleState(dvrInUse: boolean) {
    if (dvrInUse) {
      this.$el.find('#gplayer-mc-back-to-live').show()
      this.$el.find('#gplayer-mc-live').hide()
    } else {
      this.$el.find('#gplayer-mc-back-to-live').hide()
      this.$el.find('#gplayer-mc-live').show()
    }
  }
}
