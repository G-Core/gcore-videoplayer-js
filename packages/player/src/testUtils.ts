import { $, UICorePlugin } from '@clappr/core'
import Events from 'eventemitter3'
import { vi } from 'vitest'

export function createMockCore(
  options: Record<string, unknown> = {},
  container: any = createMockContainer(options),
) {
  const el = document.createElement('div')
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    $el: $(el),
    activePlayback: container.playback,
    activeContainer: container,
    i18n: {
      t: vi.fn().mockImplementation((key: string) => key),
    },
    options: {
      ...options,
    },
    configure: vi.fn(),
    getPlaybackType: vi.fn(),
    getPlugin: vi.fn(),
    load: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockPlugin() {
  return Object.assign(new Events(), {
    enable: vi.fn(),
    disable: vi.fn(),
  })
}

export function createSpinnerPlugin() {
  return Object.assign(createMockPlugin(), {
    show: vi.fn(),
    hide: vi.fn(),
  })
}

export function createMockPlayback(name = 'mock') {
  const emitter = new Events()
  return Object.assign(emitter, {
    name,
    currentLevel: -1,
    el: document.createElement('video'),
    dvrEnabled: false,
    dvrInUse: false,
    isAudioOnly: false,
    levels: [],
    consent: vi.fn(),
    play: vi.fn(),
    pause: vi.fn(),
    stop: vi.fn(),
    destroy: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    getDuration: vi.fn().mockImplementation(() => 100),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockImplementation(() => 0),
    getCurrentTime: vi.fn().mockImplementation(() => 0),
    isHighDefinitionInUse: vi.fn().mockImplementation(() => false),
    mute: vi.fn(),
    unmute: vi.fn(),
    volume: vi.fn(),
    configure: vi.fn(),
    attemptAutoPlay: vi.fn().mockImplementation(() => true),
    canAutoPlay: vi.fn().mockImplementation(() => true),
    onResize: vi.fn().mockImplementation(() => true),
    setPlaybackRate: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockContainer(
  options: Record<string, unknown> = {},
  playback: any = createMockPlayback(),
) {
  const el = playback.el
  const emitter = new Events()
  return Object.assign(emitter, {
    el,
    playback,
    options: {
      ...options,
    },
    $el: $(el),
    disableMediaControl: vi.fn(),
    enableMediaControl: vi.fn(),
    enterPiP: vi.fn(),
    exitPiP: vi.fn(),
    getDuration: vi.fn().mockReturnValue(0),
    getPlugin: vi.fn(),
    getPlaybackType: vi.fn(),
    getStartTimeOffset: vi.fn().mockReturnValue(0),
    isDvrInUse: vi.fn().mockReturnValue(false),
    isDvrEnabled: vi.fn().mockReturnValue(false),
    isHighDefinitionInUse: vi.fn().mockReturnValue(false),
    isPlaying: vi.fn().mockReturnValue(false),
    play: vi.fn(),
    seek: vi.fn(),
    seekPercentage: vi.fn(),
    switchAudioTrack: vi.fn(),
    trigger: emitter.emit,
  })
}

export function createMockMediaControl(core: any) {
  const mediaControl = new UICorePlugin(core)
  // TODO <div class="media-control-layer">
  mediaControl.$el.html(
    `<div class="media-control-left-panel" data-media-control></div>
    <div class="media-control-right-panel" data-media-control></div>
    <div class="media-control-center-panel" data-media-control></div>`,
  )
  // @ts-ignore
  mediaControl.mount = vi.fn()
  // @ts-ignore
  mediaControl.container = core.activeContainer
  // @ts-ignore
  mediaControl.toggleElement = vi.fn()
  vi.spyOn(mediaControl, 'trigger')
  core.$el.append(mediaControl.$el)
  return mediaControl
}

export function createMockBottomGear(core: any) {
  const plugin: any = new UICorePlugin(core)
  plugin.getItem = vi.fn()
  plugin.addItem = vi.fn().mockImplementation((name: string, $el: any) => {
    const existing = plugin.$el.find(`[data-${name}]`)
    if (existing.length) {
      return existing
    }
    return $('<li></li>')
      .attr(`data-${name}`, '')
      .append($el)
      .appendTo(plugin.$el)
  })
  plugin.refresh = vi.fn()
  return plugin
}
