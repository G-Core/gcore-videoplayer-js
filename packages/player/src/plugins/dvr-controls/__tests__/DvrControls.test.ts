import { beforeEach, describe, expect, it, vi } from 'vitest'
import { DvrControls } from '../DvrControls.js'
import { createMockCore, createMockMediaControl } from '../../../testUtils.js'
import { Events, Playback } from '@clappr/core'
// import { LogTracer, Logger, setTracer } from '@gcorevideo/utils'

// setTracer(new LogTracer('DvrControls.test'))
// Logger.enable('*')

describe('DvrControls', () => {
  let core: any
  let mediaControl: any
  let plugins: Record<string, any> = {}
  let dvrControls: DvrControls
  beforeEach(() => {
    core = createMockCore()
    mediaControl = createMockMediaControl(core)
    plugins = {
      media_control: mediaControl,
    }
    core.getPlugin.mockImplementation((name: string) => plugins[name])
    dvrControls = new DvrControls(core)
    core.emit(Events.CORE_READY)
    mediaControl.trigger(Events.MEDIACONTROL_RENDERED)
    core.trigger(Events.CORE_ACTIVE_CONTAINER_CHANGED, core.activeContainer)
  })
  describe('basically', () => {
    it('should render', () => {
      expect(dvrControls.el.innerHTML).toMatchSnapshot()
      expect(dvrControls.el.textContent).toMatch(/\blive\b/)
      expect(dvrControls.el.textContent).toMatch(/\bback_to_live\b/)
    })
  })
  describe('while playback type is unknown', () => {
    it('should not mount', () => {
      expect(mediaControl.slot).not.toHaveBeenCalledWith('dvr', dvrControls.$el)
    })
  })
  describe('live stream', () => {
    beforeEach(() => {
      core.getPlaybackType.mockReturnValue('live')
      core.activeContainer.getPlaybackType.mockReturnValue('live')
      core.activePlayback.getPlaybackType.mockReturnValue('live')
    })
    describe.each([
      ['no DVR', false, false, false],
      ['DVR at live edge', true, false, false],
    ])('%s', (_, dvrEnabled, dvrInUse) => {
      beforeEach(() => {
        core.activePlayback.dvrEnabled = dvrEnabled
        core.activeContainer.isDvrEnabled.mockReturnValue(dvrEnabled)
        core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
        core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
        if (dvrInUse) {
          core.activePlayback.dvrInUse = true
          core.activeContainer.isDvrInUse.mockReturnValue(true)
          core.activeContainer.emit(
            Events.CONTAINER_PLAYBACKDVRSTATECHANGED,
            true,
          )
        }
      })
      // TODO let the media control itself handle this
      it('should hide duration and position indicators', () => {
        expect(mediaControl.toggleElement).toHaveBeenCalledWith(
          'duration',
          false,
        )
        expect(mediaControl.toggleElement).toHaveBeenCalledWith(
          'position',
          false,
        )
      })
      it('should mount to the media control', () => {
        expect(mediaControl.mount).toHaveBeenCalledWith('left', dvrControls.$el)
      })
      if (dvrEnabled) {
        if (dvrInUse) {
          it('should show back_to_live button', () => {
            expect(
              dvrControls.$el.find('#gplayer-mc-back-to-live').css('display'),
            ).not.toBe('none')
          })
          it('should hide live indicator', () => {
            expect(
              dvrControls.$el.find('#gplayer-mc-live').css('display'),
            ).toBe('none')
          })
        } else {
          it('should show live indicator', () => {
            expect(
              dvrControls.$el.find('#gplayer-mc-live').css('display'),
            ).not.toBe('none')
          })
          it('should hide back_to_live button', () => {
            expect(
              dvrControls.$el.find('#gplayer-mc-back-to-live').css('display'),
            ).toBe('none')
          })
        }
      }
    })
    describe('when back_to_live button is clicked', () => {
      beforeEach(() => {
        core.activePlayback.dvrEnabled = true
        core.activeContainer.isDvrEnabled.mockReturnValue(true)
        core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
        core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
        core.activeContainer.getDuration.mockReturnValue(180)
        core.activeContainer.emit(
          Events.CONTAINER_PLAYBACKDVRSTATECHANGED,
          true,
        )
        dvrControls.$el.find('#gplayer-mc-back-to-live').click()
      })
      it('should play stream', () => {
        expect(core.activeContainer.play).toHaveBeenCalled()
      })
      it('should seek to live edge', () => {
        expect(core.activeContainer.seek).toHaveBeenCalledWith(180)
      })
    })
  })
  describe('VOD stream', () => {
    beforeEach(() => {
      core.getPlaybackType.mockReturnValue(Playback.VOD)
      core.activeContainer.getPlaybackType.mockReturnValue(Playback.VOD)
      core.activePlayback.getPlaybackType.mockReturnValue(Playback.VOD)
      core.activePlayback.emit(Events.PLAYBACK_LOADEDMETADATA)
      core.activeContainer.emit(Events.CONTAINER_LOADEDMETADATA)
    })
    // TODO handle mount points in MediaControl
    it('should not mount', () => {
      expect(mediaControl.mount).not.toHaveBeenCalledWith(
        'left',
        expect.anything(),
      )
    })
    it('should not alter position and duration indicators', () => {
      expect(mediaControl.toggleElement).not.toHaveBeenCalledWith(
        'duration',
        false,
      )
      expect(mediaControl.toggleElement).not.toHaveBeenCalledWith(
        'position',
        false,
      )
    })
  })
})
