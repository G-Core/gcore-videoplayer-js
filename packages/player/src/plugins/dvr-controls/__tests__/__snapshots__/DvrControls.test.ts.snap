// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DvrControls > basically > should render 1`] = `
"<div class="live-info" id="media-control-live">live</div>
<button type="button" class="live-button" aria-label="back_to_live" id="media-control-back-to-live">back_to_live</button>
"
`;
